lib/p2p/packets/Packet.ts

```diff
@@ -1,11 +1,43 @@
 import PacketType from './PacketType';
+import CryptoJS from 'crypto-js';
+import MD5 from 'crypto-js/md5';
+
+type PacketHeader = {
+  hash: string;
+  reqHash?: string;
+};
+
+enum MessageType {
+  UNILATERAL,
+  REQUEST,
+  RESPONSE,
+}
 
 abstract class Packet {
+  public header!: PacketHeader;
   public abstract type: PacketType;
+  public abstract messageType: MessageType;
   public abstract body: any;
-  public abstract responseType?: PacketType;
-  public abstract responseTimeout?: number;
-  public abstract toRaw(): string;
+
+  public fromRaw(raw: string): Packet {
```

Comment:
`fromRaw` as a non-static method doesn't seem right. What about making this a static method and use generics to define what class gets returned?

Reply:
Generics won’t work here, because you can’t instantiate a new object by using `T` (unlike in C#). So you will must pass the constructor to the method, which seems hacky to me. 
For the non-static way it won't work either, because a sub-class is not comparable with the base-class.

It works fine as a static method if you implement it on each sub-class, but I wanted to avoid it, because as for now the implementation is exactly the same, and plus, it gives a nice separation of serialization/deserialization from the sub-classes.

Reply:
Ahh I wasn't aware that this wasn't possible but it seems you are right.